tls/s2n_config.c

```diff
@@ -630,12 +630,24 @@ int s2n_config_set_extension_data(struct s2n_config *config, s2n_tls_extension_t
 
 int s2n_config_set_client_hello_cb(struct s2n_config *config, s2n_client_hello_fn client_hello_cb, void *ctx)
 {
+    POSIX_ENSURE_REF(config);
+
     config->client_hello_cb = client_hello_cb;
     config->client_hello_cb_ctx = ctx;
-
+    config->client_hello_cb_mode = S2N_CLIENT_HELLO_CB_BLOCKING;
```

Comment:
So if I call `s2n_config_set_client_hello_cb_mode` to set the callback as non-blocking, but then call s2n_config_set_client_hello_cb, I get a blocking callback?

I would suggest setting client_hello_cb_mode to a default during s2n_config_new, not when setting the callback. We try to avoid configuration API order affecting behavior.

Reply:
nice catch. Moved to `s2n_config_init()`